core/modules/replica/HealthMonitorThread.cc

```diff
@@ -0,0 +1,233 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/HealthMonitorThread.h"
+
+// System headers
+#include <atomic>
+#include <map>
+
+// Qserv headers
+#include "replica/ClusterHealthJob.h"
+#include "util/BlockPost.h"
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+HealthMonitorThread::Ptr HealthMonitorThread::create(Controller::Ptr const& controller,
+                                                     ControlThread::CallbackType const& onTerminated,
+                                                     WorkerEvictCallbackType const& onWorkerEvictTimeout,
+                                                     unsigned int workerEvictTimeoutSec,
+                                                     unsigned int workerResponseTimeoutSec,
+                                                     unsigned int healthProbeIntervalSec) {
+    return Ptr(
+        new HealthMonitorThread(
+            controller,
+            onTerminated,
+            onWorkerEvictTimeout,
+            workerEvictTimeoutSec,
+            workerResponseTimeoutSec,
+            healthProbeIntervalSec
+        )
+    );
+}
+
+HealthMonitorThread::WorkerResponseDelay HealthMonitorThread::workerResponseDelay() const {
+    util::Lock lock(_mtx, "HealthMonitorThread::workerResponseDelay()");
+    return _workerServiceNoResponseSec;
+}
+
+void HealthMonitorThread::run() {
+ 
+    std::string const context = "HealthMonitorThread::run()";
+    {
+        util::Lock lock(_mtx, context);
+
+        for (auto&& worker: serviceProvider()->config()->workers()) {
+            _workerServiceNoResponseSec[worker]["qserv"] = 0;
+            _workerServiceNoResponseSec[worker]["replication"] = 0;
+        }
+    }
+
+    std::string const parentJobId;  // no parent jobs
+
+    while (not stopRequested()) {
+
+        // Probe hosts. Wait for completion or expiration of the job
+        // before analyzing its findings.
+
+        info("ClusterHealthJob");
+
+        std::atomic<size_t> numFinishedJobs{0};
+        std::vector<ClusterHealthJob::Ptr> jobs;
+        jobs.emplace_back(
+            ClusterHealthJob::create(
+                _workerResponseTimeoutSec,
+                controller(),
+                parentJobId,
+                [&numFinishedJobs](ClusterHealthJob::Ptr const& job) {
+                    ++numFinishedJobs;
```

Comment:
it seems awfully brave to me to be sharing a function variable with other threads. I see that this is the run function, so I think I can see it is the run loop of a thread, which maybe makes it better. But still, what if this thread exits, perhaps unexpectedly?
One alternative I can suggest is: in the lambda capture a `weak_ptr` to `this`, and have the body of the function lock the weak ptr and if it succeeds, call a function `incrementFinishedJobs()` (or increment the atomic var directly but I somewhat prefer the accessor function).

Reply:
Thanks! That's the correct observation (regarding `numFinishedJobs`). I will reinforce the code to pass a shared pointer to `HealthMonitorThread` instead. This will guarantee the lifespan of `numFinishedJobs` for as long as an instance of `ClusterHealthJob` is around.